### PR TITLE
[SPARK-37686][PYTHON][SQL] Use _invoke_function helpers for all pyspark.sql.functions

### PR DESCRIPTION
**Pull request by zero323**
### What changes were proposed in this pull request?

This PR proposes conversion of functions not covered by SPARK-32084 to `_invoke_functions` style.

Two new `_invoke` functions where added:

- `_invoke_function_over_columns`
- `_invoke_function_over_seq_of_columns`

to address common examples.

### Why are the changes needed?

To reduce boilerplate (especially related to type checking) and improve manageability.

Additionally, it opens opportunity for reducing driver-side invocation latency.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Existing tests.

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
Wow .. this is a big change. cc @ueshin @BryanCutler @viirya too FYI. I will review this closely within this week 👍

**Comment by zero323** [comment]
> Wow .. this is a big change. cc @ueshin @BryanCutler @viirya too FYI. I will review this closely within this week 👍

Just to make the context clear so it doesn't look like a flimsy ‒ this alone is nice to have IMHO, because `pyspark.sql.functions` are the fastest growing module and adding new functions should require as little effort as possible.

But there is another opportunity here ‒ with this, and few minor additional tweaks, we can can cache `_get_jvm_function`

```python
from functools import lru_cache


@lru_cache(192)
def _get_jvm_function(name: str, sc: SparkContext) -> Callable:
   ...
```
 
significantly reducing the overhead of invoking the whole thing.

In simple experiments this:

```python
>>> %%timeit -n 100
... for i in range(1000): col(f"a{i}")
...
```

currently takes

```
1.08 s ± 9.44 ms per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

and after caching function retrieval:

```
116 ms ± 3.73 ms per loop (mean ± std. dev. of 7 runs, 100 loops each)
```


Maybe not a big deal for batch applications, but if you have complex query in `foreachBatch`, it is a nice gain (and it is still around order of magnitude difference).

File: python/pyspark/sql/functions.py [hunk 1]
```diff
@@ -1264,10 +1269,7 @@ def grouping(col: "ColumnOrName") -> Column:
     |  Bob|             0|       5|
     +-----+--------------+--------+
     """
-    sc = SparkContext._active_spark_context
-    assert sc is not None and sc._jvm is not None
-    jc = sc._jvm.functions.grouping(_to_java_column(col))
-    return Column(jc)
+    return _invoke_function_over_column("grouping", col)
```
Comment (ueshin) [inline python/pyspark/sql/functions.py:1272]: ditto, and following several invocations?

File: python/pyspark/sql/functions.py [hunk 2]
```diff
@@ -79,16 +80,34 @@ def _invoke_function(name: str, *args: Any) -> Column:
...
 
+def _invoke_function_over_columns(name: str, *cols: "ColumnOrName") -> Column:
+    """
+    Invokes n-ary JVM function identified by name
+    and wraps the result with :class:`~pyspark.sql.Column`.
+    """
+    return _invoke_function(name, *(_to_java_column(col) for col in cols))
+
+
 def _invoke_function_over_column(name: str, col: "ColumnOrName") -> Column:
```
[the remark this answers is not in the document]
Comment (zero323) [reply python/pyspark/sql/functions.py:95]: That was my initial choice and then I looked at  all the `_invoke_function_over_columns(col)` and it seemed a little bit confusing ‒ like there is something missing. It is highly subjective though, so I won't insist on this particular approach :)

File: python/pyspark/sql/functions.py [hunk 3]
```diff
@@ -1630,10 +1593,7 @@ def shiftrightunsigned(col: "ColumnOrName", numBits: int) -> Column:
     >>> df.select(shiftrightunsigned('a', 1).alias('r')).collect()
     [Row(r=9223372036854775787)]
     """
-    sc = SparkContext._active_spark_context
-    assert sc is not None and sc._jvm is not None
-    jc = sc._jvm.functions.shiftRightUnsigned(_to_java_column(col), numBits)
-    return Column(jc)
+    return _invoke_function("shiftRightUnsigned", _to_java_column(col), numBits)
```
Comment (viirya) [inline python/pyspark/sql/functions.py:1596]: shiftrightunsigned?

**Review by viirya — APPROVED** [review]

**Comment by viirya** [comment]
Looks okay.

**Review by ueshin — APPROVED** [review]

**Comment by zero323** [comment]
Merged into master.

Thanks everyone!

**Pull request closed without merging**